TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs

```diff
@@ -1,192 +1,232 @@
+            var responseMessage = actionResult.ExecuteAsync(CancellationToken.None).Result;
+            HttpError error;
+            responseMessage.TryGetContentValue(out error);
+            var actualLocation = responseMessage.Headers.Location;
+
+            Assert.That(responseMessage.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest));
+            Assert.That(error.ModelState.Keys, Is.EqualTo(expectedKeys));
+            Assert.That(actualLocation, Is.EqualTo(null));
+        }
+
+        [Test]
+        public void PostAsync_WithValidArguments_ReturnsCorrectResponse()
+        {
+            var expectedListItem = new ListItem { Id = _guidOfFirstItem, Text = "text" };
+            var postedListItem = new ListItem { Id = Guid.Empty, Text = "newText" };
+            var expectedLocation = new Uri($"api/v1/ListItems/{_guidOfFirstItem}", UriKind.Relative);
+            _itemCreationService.CreateNewItemAsync(postedListItem).Returns(expectedListItem);
+            _urlGenerator.GenerateUrl(expectedListItem).Returns(callInfo => $"api/v1/ListItems/{callInfo.Arg<ListItem>().Id}");
+
+            var actionResult = _controller.PostAsync(postedListItem).Result;
+            var responseMessage = actionResult.ExecuteAsync(CancellationToken.None).Result;
+            var actualLocation = responseMessage.Headers.Location;
+            ListItem actualListItem;
+            responseMessage.TryGetContentValue(out actualListItem);
+
+            Assert.That(responseMessage.StatusCode, Is.EqualTo(HttpStatusCode.Created));
+            Assert.That(actualListItem, Is.EqualTo(expectedListItem).UsingListItemComparer());
+            Assert.That(actualLocation, Is.EqualTo(expectedLocation));
+        }
+
+        [Test]
+        public void PutAsync_WithValidArguments_ReturnsCorrectResponse()
+        {
+            var databaseListItem = new ListItem { Id = _guidOfThirdItem, Text = "random" };
+            var expectedListItem = new ListItem { Id = _guidOfThirdItem, Text = "updated" };
+            var updatedListItem = new ListItem { Id = _guidOfThirdItem, Text = "newText" };
+            var acquisitionResult = AcquisitionResult.Create(databaseListItem);
+            _itemAcquisitionService.GetItemAsync(_guidOfThirdItem).Returns(acquisitionResult);
+            _itemModificationService.UpdateExistingItemAsync(acquisitionResult, updatedListItem).Returns(expectedListItem);
+
+            var actionResult = _controller.PutAsync(updatedListItem).Result;
+            var responseMessage = actionResult.ExecuteAsync(CancellationToken.None).Result;
+            ListItem actualListItem;
+            responseMessage.TryGetContentValue(out actualListItem);
+
+            Assert.That(responseMessage.StatusCode, Is.EqualTo(HttpStatusCode.OK));
+            Assert.That(actualListItem, Is.EqualTo(expectedListItem).UsingListItemComparer());
+        }
+
+        [Test]
+        public void PutAsync_ItemWithNonExistingId_ReturnsCorrectErrorResponse()
```

Comment:
This test is a good addition to the one with "happy path", however, I miss tests for malformed request (where `ListItem item` argument could not be parsed by the framework) and for the fast-failure check for well known unsupported `Guid` values (→ `item.Id == Guid.Empty`)